gr-filter: Add ichar and ishort decimator

ghostop14 opened the pull request:
Current resamplers do not support ichar as an input type.  However,
when processing high-speed wide bandwidth streaming ichar 
data from external sources (such as the ATA), decimation in 
native ichar would be faster than converting
to another data type then decimating.  This block provides the ability
to decimate the ichar stream directly.

mormj commented:
@ghostop14 - if this block is doing straight decimation with no filtering as it appears to be doing, shouldn't we have that functionality for all input types?  Seem like it would be inconsistent in the block set to have a signal processing function specific to one data type.  So - could this be just a decimator block that takes in size_t for the sizeof_input_items?

Also, could you just use keep 1 in N?

michaelld commented on gr-filter/include/gnuradio/filter/ichar_decimator.h:
```diff
@@ -0,0 +1,41 @@
+public:
+    typedef boost::shared_ptr<ichar_decimator> sptr;
+
+
```
whitespace

michaelld commented on gr-filter/lib/ichar_decimator_impl.cc:
```diff
@@ -0,0 +1,71 @@
+    d_doubledecimation = 2 * decimation;
+
+    // Make sure we work with pairs of bytes (I and Q as byte)
+    gr::block::set_output_multiple(2);
```
should this be `d_doubledecimation` instead, to make sure in `work` that there are an integer number of decimations? Otherwise I believe that there could be glitches where more or fewer than `decimation` complex<char> are dropped.

michaelld commented:
I'm wondering the same thing as @mormj  ... this seems to be a "keep 1 in N" block ... pre-filtering required ... so what makes this block special? I do like the GRC scripts & interface!

ghostop14 commented:
Hi @michaelld and @mormj, fair questions.  A little history may help.  But first, @michaelld in answer to the set_output_items, I think 2 is still the correct answer.  If you use doubledecimation as the multiplier, increased decimation would increase output multiples.  Because the input and output counters are tracked separately and the actual number of outputs is returned from work rather than returning noutput_items, no samples should be missed (any unused should be sent in on the next work call).

For the justification for the block, a little history: The complex/ichar scaling (#3208), this block, and a good bit of refactoring in grnet that I'll have time to focus on a PR for the first week in March all came out of work done at the ATA hackathon last year.  The metric was 104.8576 MCS/s, 2 feeds: H and V polarization for a total of about 3.354 Gbps of streaming ichar data from the beamformer.  So every drop of performance counted.  Folks doing 10 MCSPS wouldn't notice it but at those speeds it mattered to not drop packets.  #3204 allowed for scaling the voltage-based ichar inputs, and this block saved the memcpy overhead of having to go stream-to-vector:keep 1-in-n:vector-to-stream which necessitated 2 extra memcpy's, 2 extra blocks/threads, and a tag loop.  So this block does it at the highest performance possible in a single loop to keep up with the stream.  I thought about enhancing the rational resampler which doesn't currently support ichar or byte, but the fact that it also takes filter taps felt like it wasn't the right spot for this functionality either and that's what drove this to be a new/unique block.  Right now the gr-ata flowgraphs require gr-grnet, but with these 2 PR's then the integration of the net blocks in March, gr-ata would be native GR 3.9.

Thoughts?

michaelld commented:
> But first, @michaelld in answer to the set_output_items, I think 2 is still the correct answer. If you use doubledecimation as the multiplier, increased decimation would increase output multiples. Because the input and output counters are tracked separately and the actual number of outputs is returned from work rather than returning noutput_items, no samples should be missed (any unused should be sent in on the next work call).

After reviewing the GR classes `sync_decimator` and `sync_block`, I agree with you here. I hadn't put all of the pieces together. Well done!

michaelld commented:
@ghostop14 so the backstory is: this block -is- keep 1 in N, but is separate for performance reasons. hmmm ... no tags here ... should tags be considered? maybe an option that the user can decline? You could consider rewriting `keep_one_in_n_impl` to be more efficient ... it's a nice academic implementation, which could be a lot faster!

michaelld commented:
Also wondering if there should be `ishort` & `iint` versions ...

ghostop14 commented:
I think it would be pretty easy to add ishort and iint versions.  Could be nice coming out of a USRP block.  Maybe I rename it IVal Decimator and give it drop-options for ishort, iint, and ichar.   On the tags, I always hate the extra get_tags and additional for loop when performance is key, but I think we could do that with a block configuration option and just if-check it in the work() if you think folks would really want tag capabilities too?

ghostop14 commented:
@michaelld, renamed it to ival decimator and added support for ishorts in the same way as a drop-list option.  I initially added iint but it didn't look like we had other blocks that supported Complex->int / int->complex in the same way, just ichar and ishort, so I removed it.

I don't know the answer to this so this is a question.... looking at Keep-1-in-n it mentions it only has to manually deal with tags since it sets a TPP_DONT policy.  Since we're not doing that here (and I think this is the case), tags are automatically taken care of and we don't need to manually do it in the work function since we're not setting that policy?  Is that accurate?

michaelld commented:
> @michaelld, renamed it to ival decimator and added support for ishorts in the same way as a drop-list option. I initially added iint but it didn't look like we had other blocks that supported Complex->int / int->complex in the same way, just ichar and ishort, so I removed it.

Sounds good.

> I don't know the answer to this so this is a question.... looking at Keep-1-in-n it mentions it only has to manually deal with tags since it sets a TPP_DONT policy. Since we're not doing that here (and I think this is the case), tags are automatically taken care of and we don't need to manually do it in the work function since we're not setting that policy? Is that accurate?

Hmm ... I'll need to look at the core code that handles tags. I'm guessing there's a reason why the K1inN does tags the way it does ... just a guess though until I have hard facts from a code review.

michaelld commented:
> On the tags, I always hate the extra get_tags and additional for loop when performance is key, but I think we could do that with a block configuration option and just if-check it in the work() if you think folks would really want tag capabilities too?

I do think folks expect and also want tags to propagate properly through most blocks. Hence, I could see this as an option: propagate tags properly or don't at all. Default would be don't.

ghostop14 commented:
Sounds good.  Let me know what you think from the core code if they propagate automatically all things being equal.  If its not handled automatically I’ll add it.

michaelld commented on gr-filter/lib/ival_decimator_impl.cc:
```diff
@@ -0,0 +1,98 @@
+{
+    d_doubledecimation = 2 * decimation;
+
+    if ((d_data_type < 1) || d_data_type > 2)
```
need a check for `data_size` too

michaelld replied:
can you just use `data_type` for both it and `data_size` ... since they happen to overlap exactly?

michaelld commented on gr-filter/include/gnuradio/filter/ival_decimator.h:
```diff
@@ -0,0 +1,41 @@
+public:
+    typedef boost::shared_ptr<ival_decimator> sptr;
+
+
```
Extract space

michaelld commented on gr-filter/grc/filter.tree.yml:
```diff
@@ -26,6 +26,7 @@
   - pfb_arb_resampler_xxx
   - rational_resampler_xxx
   - rational_resampler_base_xxx
+  - ival_decim
```
Is this supposed to be the block name or the file name? I'm still not sure ...

ghostop14 commented:
@michaeld: Fixed the header and added size checks.  I don't think I want to just rely on size for data type checks.  If in the future we decide that int / float makes sense we can run into scenarios where data sizes are the same for multiple data types.  For the tree.yml, it's a good question.  I made them all match just to be safe.

michaelld commented on gr-filter/lib/ival_decimator_impl.cc:
```diff
@@ -0,0 +1,110 @@
+ */
+ival_decimator_impl::ival_decimator_impl(int decimation, int data_type, int data_size)
+    : gr::sync_decimator("ival_decimator",
+                         gr::io_signature::make(1, 1, data_size),
```
Since `data_size` and `data_type` -are- directly related, let's get rid of 1 of them. either one will work alone; using both is redundant and prone to user error.

mormj replied:
Why wouldn't we just use data_size so it opens the block up to all data types.  Not a big fan of the new #defines.  If we go with data_type, then we should probably be able to reuse the gr enums that are exposed through swig.  With just data_size (like other blocks), could the if statement be removed and void *'s be std::copy'd two elements at a time?

michaelld replied:
Using a `data_type` allows the `switch` and then using normal C/C++ types to do the work, which will be a -lot- more efficient than using `memcpy` or whatever. That said, the reverse could be done too: use the correct `int` type for the size provided, then do the style of copying as done already. Since this block isn't -interpreting- the data, just moving the value around, this would work.

michaelld replied:
NOTE: of course this wouldn't work for 3 byte items, but we don't use those in GR anyway. No sure how to handle 16 byte items, since there is no `int128_t` yet TTBOMK ... but everything up to 8 byte items would be doable.

michaelld commented on gr-filter/include/gnuradio/filter/ival_decimator.h:
```diff
@@ -0,0 +1,40 @@
+ * \ingroup resamplers_blk
+ *
+ * \details
+ * This block provides interleaved char decimation without
```
please correct this comment, meaning "`char` or `short`" ... or something that represents the new block data types

michaelld commented:
Asking for at least 1 more review of this PR, since it's about ready to go.

ghostop14 commented:
Comment fixed.

mormj commented:
> 
> 
> Also wondering if there should be `ishort` & `iint` versions ...

I'm still hoping there's a way to do it without new #defines and so it isn't an if or switch for every type.

Also - can the same performance improvements be gained from non-interleaved types, so that this just becomes a vanilla decimator block?  gr_complex, float, etc.  Or will these just have the net effect of keep 1 in N

mormj commented on gr-filter/grc/filter_ival_decimator.block.yml:
```diff
@@ -0,0 +1,34 @@
+id: ival_decimator
+label: IValue Stream Decimator
```
If this is strictly for i values - perhaps the label should read "Interleaved Stream Decimator" to spell it out

mormj commented on gr-filter/grc/filter_ival_decimator.block.yml:
```diff
@@ -0,0 +1,34 @@
+    dtype: enum
+    options: [byte, short]
+    option_attributes:
+        datatype: ['1', '2']
```
As noted in the other comments, I'm not a fan of the hardcoded mappings here - would be nice to use existing enums, or just use the sizeof

ghostop14 commented:
Okay folks, a couple updates just pushed and some perspectives too on design choices I made.  So I changed the #defines to constexpr's.  I may be old-school but in this case I think #defines were fine.  But I get the impression gr is trying to clean them out of the code, so they're constexpr's now.  I changed the name to "Interleaved Stream Operator" as requested as well.  I'm concerned it's long on a flowgraph which is why I had kept it short (I hate blocks that take up a lot of horizontal space just for the name.

In terms of the size versus type, trust me, this is easier to scale later.  And because data size isn't unique it can cause problems and unintended redo's later.  I personally prefer the strongly typed switch case blocks in work over memcpy's.  Same scenario as moving from #defines to constexpr's.  Take advantage of the strong types.  It's less error-prone than the memcpy's.  Just my design preferences, but I'm happy to consider better ways.  

Just FYI, I'm going to be short on time to work on these till the beginning of March after today so I'd like to try to wrap them up so they don't sit if possible.

michaelld commented:
Thanks for fixing the comment. I'm fine with the PR as is. Unless someone is completely against the way the `data_type` argument is set, I'd like to get this merged & then we can revisit it down the road if/as necessary. The performance will be a -lot- better than K-1-i-N, though not as robust overall -- a typical trade-off we make in programming!

michaelld commented:
I'd guess that many blocks could use a `data_type` concept to help speed them up! We're not going to cover that here in this PR today :)

ghostop14 commented:
After several side conversations, I did remove the data_type in the interest of integration, so have another look.  Just for the record I still stand by the other approach as the better design for future expansion.  In my experience, choosing a non-unique identifier will come back to bite you down the road when you want to add to it.  But in the interest of getting it merged, I changed it to rely on size only.

michaelld commented on gr-filter/lib/ival_decimator_impl.cc:
```diff
@@ -0,0 +1,90 @@
+      d_data_size(data_size)
+{
+    d_doubledecimation = 2 * decimation;
+
```
please add a check for `data_size` to be either `sizeof(char)` or `sizeof(short)` ... or throw an error here during instantiation.

mormj commented on gr-filter/grc/filter_ival_decimator.block.yml:
```diff
@@ -0,0 +1,34 @@
+    dtype: enum
+    options: [byte, short]
+    option_attributes:
+        datasize: [gr.sizeof_char, gr.sizeof_short]
```
This makes me feel much better with enums

ghostop14 commented:
I just pushed the update to address the data validation.  Thanks for all the reviews!  I know it's time-consuming.

michaelld commented on gr-filter/lib/ival_decimator_impl.cc:
```diff
@@ -0,0 +1,93 @@
+      d_data_size(data_size)
+{
+    if ((data_size != 1) && (data_size != 2))
+        throw std::invalid_argument("data size must be in 1 (char) or 2 (short)");
```
yes!

michaelld commented:
CI passes. 2 maintainer approvals. Merging.
